### PR TITLE
Codex/ios cjk input fix

### PR DESCRIPTION
**Pull request by fufesou**
Original PR https://github.com/rustdesk/rustdesk/pull/15072 , but the author has not replied to the comments. This PR is filed as a replacement.

Fixes iOS soft keyboard handling for CJK IMEs in remote sessions.

The previous iOS text diff logic treated any active composing range as text that should not be sent. Some iOS IMEs keep a composing range briefly after the user confirms a candidate, so committed Chinese text could be held until another key, such as space, causes another text change.

## Changes

- Extracted iOS soft keyboard diffing into `ios_soft_keyboard_input.dart`.
- Tracks pending composing text so pinyin, stroke, bopomofo, Japanese kana, and Korean jamo/hangul composition are not sent prematurely.
- Sends committed CJK text immediately via `sessionInputString`, including Japanese kanji/kana and Korean hangul once composition is committed.
- Uses a text controller listener on iOS so commits where only the composing range changes are still observed.
- Added focused tests for Chinese, Japanese, Korean, ASCII, and backspace behavior.

### Japanese

https://github.com/user-attachments/assets/9f4bbfab-a2f8-4b86-abbd-a953d8bdb11f

https://github.com/user-attachments/assets/e14353be-98db-4314-baf7-0c65480efe24

### Korean

https://github.com/user-attachments/assets/03cf8642-1c43-4e86-8b4f-28ce53f92aa7

https://github.com/user-attachments/assets/e854f13b-54f0-42e6-9001-a4c796ac0199

### Chinese Pinyin

https://github.com/user-attachments/assets/7a7a671e-e1c4-4990-a68e-d2968c0a5923

https://github.com/user-attachments/assets/9c7cc05d-c70a-4416-8a0c-d9f83506e1c9

### Chinese Zhuyin

https://github.com/user-attachments/assets/aec7639e-b51f-439c-9fb2-c2f542ca9153

https://github.com/user-attachments/assets/32a45822-3f89-4349-859a-0520caa50523

## Testing

- [x] iOS soft keyboard basics: ASCII, space, return, backspace
- [x] iOS CJK IME: Korean, Chinese pinyin/stroke/bopomofo, Japanese
- [x] iOS accents and emoji smoke tests
- [x] Android soft keyboard basics: ASCII, space, return, backspace
- [x] `flutter test --no-pub test/ios_soft_keyboard_input_test.dart test/input_modifier_utils_test.dart`
- The Microsoft Swift keyboard does not work. The master branch also does not work.

## Known limitation:

- handwriting and Korean input can still expose intermediate composing text on the controlled side; this is left as-is for now because holding those events caused worse regressions

## Known issues

- Pressing multiple keys simultaneously, such as "d" and "f", will select the hidden TextFormField input, causing the Sentinel's baseline to shift. Subsequent input may result in the controlled text continuously deleting a large chunk of content, continuously inserting a large number of 1s, or deleting and then inserting.

https://github.com/user-attachments/assets/7bbc78d1-90ac-45df-9201-37235fe90fb8

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved iOS soft-keyboard composition handling for more accurate commits, deletions and emoji/rune boundaries across diverse input methods.
  * Refined keyboard visibility and edit-mode behavior to avoid stray input and better route keyboard events when the iOS soft keyboard is active, including when edit mode is shown or hidden.

* **Tests**
  * Added comprehensive tests covering iOS soft-keyboard diffing and commit/hold semantics across many composition edge cases, scripts, and sentinel-prefix scenarios.

### CODE REVIEW

**Comment by rustdesk** [comment]
• The iOS diff logic still loses valid user input in a realistic sequence involving remote backspace followed by typing the sentinel character. This breaks soft-keyboard input correctness for 1 after the
  sentinel prefix has been shortened.

  Review comment:

  - [P2] Do not swallow 1 after a shortened sentinel prefix — /Users/hzhou/rustdesk/flutter/lib/mobile/utils/ios_soft_keyboard_input.dart:232-233
    When the user sends Backspace while the local tail is empty, _value is shortened from the sentinel prefix (for example 111 -> 11). If the next real key is 1, this branch treats 11 -> 111 as a sentinel-
    prefix restore and emits no action, so the first 1 after each such remote backspace is never sent to the controlled side.

**Comment by fufesou** [comment]
@coderabbitai review

File: flutter/lib/mobile/utils/ios_soft_keyboard_input.dart [hunk 1]
```diff
@@ -0,0 +1,948 @@
...
+      String.fromCharCode(rune).trim().isEmpty;
+}
+
+Iterable<int> _runeBoundaryOffsets(String value) sync* {
+  var offset = 0;
+  for (final rune in value.runes) {
+    offset += String.fromCharCode(rune).length;
+    if (offset < value.length) yield offset;
+  }
+}
```
[the remark this answers is not in the document]
Comment (fufesou) [reply flutter/lib/mobile/utils/ios_soft_keyboard_input.dart:711]: I checked this against Dart's behavior. `String.fromCharCode(0x1F600).length` returns 2, same as the actual emoji string length, so it does not truncate supplementary code points to one UTF-16 code unit.

Because of that, the current offset calculation is already correct. The suggested change would be equivalent, but it is not needed as a fix.

For the test helper note: agreed it could be added, but the collapsed composing path is already covered directly by an explicit test, so I’d prefer to keep the helper minimal for now.

**Comment by fufesou** [comment]
@coderabbitai review

**Comment by fufesou** [comment]
@coderabbitai review

**Comment by fufesou** [comment]
@coderabbitai review

**Comment by fufesou** [comment]
@coderabbitai review

**Comment by fufesou** [comment]
@coderabbitai review

**Comment by fufesou** [comment]
ref  https://github.com/rustdesk/rustdesk/pull/15429

**Comment by Mmx233** [comment]
Sad to see this PR closed. Is there anything I can do to help move the CJK fix forward?

**Comment by fufesou** [comment]
A new PR is welcome; you can continue with this PR or rewrite the implementation.

However, testing CJK input may be tricky, so please test and verify it thoroughly.

**Comment by rustdesk** [comment]
I routinely cleaned up some long-standing draft PRs.
